### PR TITLE
GHA/windows: enable libssh in !ssl MSVC job

### PR DESCRIPTION
**Pull request by talregev**
GHA/windows: add MSVC libssh to openssl job

### CODE REVIEW

**Comment by vszakats** [comment]
Can we add libssh to an existing job?

**Comment by talregev** [comment]
> Can we add libssh to an existing job?

Yes. which one you suggest?

**Comment by vszakats** [comment]
What about the !ssl one? Or openssl uwp if libssh supports it.

Notice however that libssh has an pretty huge vulnerability on Windows, so this is a dependency nobody should use in a serious or any setting. Someone with a Gitlab account shall probably report it.

**Comment by talregev** [comment]
> What about the !ssl one? Or openssl uwp if libssh supports it.
> 
> Notice however that libssh has an pretty huge vulnerability on Windows, so this is a dependency nobody should use in a serious or any setting. Someone with a Gitlab account shall probably report it: https://github.com/curl/curl-for-win/blob/3951808deb04df9489ee17430f236ed54436f81a/libssh.sh#L6-L8 (the issue was there last time I checked a few weeks ago)

So maybe we should wait for it? 
I just added it to activate more curl tests ...

**Comment by vszakats** [comment]
In the `!ssl` job it would? But, SSH tests are not running on Windows at all, so it remains a build-only test:
```
TESTINFO: "failed starting SSH server" 51 times (582, 583, 600, 601, 602, 603, 604, 605, 606 and 42 more)
```

Building it still useful, but perhaps with a comment telling about this caveat?

/off
This is seems to be the problem (couldn't figure out how to make permalinks with GitLab, sorry):
https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/config.c#L588 `ssh_config_make_absolute()`
is loading the config from a hard-coded `/etc/ssh/`, `~/.ssh` and other unixy things. On Windows this is `C:/etc/ssh`, which is world-writable / unprotected.
Other unixy parts:
https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/misc.c#L1175 `ssh_path_expand_escape()`
https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/options.c#L1815 `ssh_options_parse_config()`
This latter function is called from `ssh_connect()` with a default (`NULL`) filename on all platforms.
A `GLOBAL_CLIENT_CONFIG` settings is configurable via CMake, and defaults to `/etc/ssh/ssh_config` on all platforms. On Windows this is effectively `<current drive>:/etc/ssh/ssh_config`, world-writable. This also cannot be configured securely, because there is no universally secure path on Windows.

**Comment by talregev** [comment]
where we should add this comment?

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -530,12 +530,12 @@ jobs:
...
           config: '-DENABLE_DEBUG=ON -DENABLE_UNICODE=OFF -DCURL_USE_SCHANNEL=OFF -DCURL_BROTLI=ON -DCURL_ZSTD=ON -DCURL_USE_LIBPSL=ON -DUSE_NGHTTP2=ON -DBUILD_SHARED_LIBS=OFF -DCURL_USE_LIBSSH2=ON -DCURL_USE_OPENSSL=ON -DUSE_HTTPSRR=ON -DUSE_ECH=ON'
         - name: '!ssl'
-          install: 'brotli zlib zstd libpsl'
+          install: 'brotli zlib zstd libpsl libssh'
           arch: 'x64'
           plat: 'windows'
           type: 'Debug'
           tflags: '~1516 ~2301 ~2302 ~2303 ~2307'
-          config: '-DENABLE_DEBUG=ON -DENABLE_UNICODE=OFF -DCURL_USE_SCHANNEL=OFF -DCURL_BROTLI=ON -DCURL_ZSTD=ON -DCURL_USE_LIBPSL=ON -DHTTP_ONLY=ON'
+          config: '-DENABLE_DEBUG=ON -DENABLE_UNICODE=OFF -DCURL_USE_SCHANNEL=OFF -DCURL_BROTLI=ON -DCURL_ZSTD=ON -DCURL_USE_LIBPSL=ON -DHTTP_ONLY=ON -DCURL_USE_LIBSSH2=OFF -DCURL_USE_LIBSSH=ON'
```
Comment (vszakats) [inline .github/workflows/windows.yml:538]: Before the `config:` line perhaps?

Comment (talregev) [reply .github/workflows/windows.yml:542]: Done.

**Comment by talregev** [comment]
> In the `!ssl` job it would? But, SSH tests are not running on Windows at all, so it remains a build-only test:
> 
> ```
> TESTINFO: "failed starting SSH server" 51 times (582, 583, 600, 601, 602, 603, 604, 605, 606 and 42 more)
> ```
> 
> Building it is still useful, but perhaps with a comment telling about this caveat?
> 
> /off This is seems to be the problem (couldn't figure out how to make permalinks with GitLab, sorry): https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/config.c#L588 `ssh_config_make_absolute()` is loading the config from a hard-coded `/etc/ssh/`, `~/.ssh`, and other unixy things. On Windows this is `C:/etc/ssh`, which is world-writable / unprotected. Other unixy parts: https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/misc.c#L1175 `ssh_path_expand_escape()` https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/options.c#L1815 `ssh_options_parse_config()` This latter function is called from `ssh_connect()` with a default (`NULL`) filename on all platforms. A `GLOBAL_CLIENT_CONFIG` setting is configurable via CMake, and defaults to `/etc/ssh/ssh_config` on all platforms. On Windows this is effectively `<current drive>:/etc/ssh/ssh_config`, world-writable. This also cannot be configured securely, because there is no universally secure path on Windows. It's almost the same issue as this two-decade old one with OpenSSL: [openssl/openssl#24528](https://github.com/openssl/openssl/issues/24528), but contrary to OpenSSL, libssh doesn't want to load DLLs via these configs.

How it done on libssh2?
Why not to read the config from .ssh in the user setting folder in windows?
`cd %USERPROFILE%\.ssh`

**Comment by vszakats** [comment]
libssh2 doesn't have an on-disk configuration. It also only reads filenames passed to it explicitly. The same class of issues can only affect the caller. In curl's case `lib/vssh/libssh2.c`. That isn't perfect (ref #13571, and lack of `USERPROFILE`) either, but it's not a libssh2 issue.

**Comment by vszakats** [comment]
>Why not to read the config from .ssh in the user setting folder in windows? `cd %USERPROFILE%\.ssh` or `cd Program Files\etc\ssh` Program files in windows is protected folder, access to every user by reading. Can access from environment variable: `cd %ProgramFiles%`

If `%USERPROFILE%` and `%ProgramFiles%` are detected at runtime, it's indeed one possible good fix. Another option is to offer an API to set these paths. Or both.

OpenSSL for example has been hard-coding `/etc/ssl` for a long time, then (after CVE-2019-5443) changed this to a _built-time_ hard-coded `%ProgramFiles%` (that is still wrong), with a buggy override and no way to disable. Lately added an option to disable some (all?) of the functionality depending on hard-coded paths. And is now shifting to a convoluted registry-based setting which requires an installer be run by an admin, with the existing bad solution still in place, for "compatibility". It's puzzling.

**Review by vszakats — APPROVED** [review]

**Comment by talregev** [comment]
%USERPROFILE% and %ProgramFiles% are detected at runtime.

**Comment by vszakats** [comment]
Can you rebase on master?

**Comment by talregev** [comment]
> Can you rebase on master?

Done.

**Comment by vszakats** [comment]
Thanks @talregev!

**Pull request closed without merging**